### ISSUE TITLE
[BUG] Vue 3 Component Reactivity not working

### ISSUE THREAD

**Issue by StepanMynarik** [microsoft/playwright#18371]
**Context:**
- Playwright Version: 1.27.1
- Operating System: Windows
- Node.js version: 16.17.0, 18.8.0
- Browser: All
- Extra: Vite, Vue 3

```playwright test -c playwright-ct.config.ts```

When creating ```ref``` with some mock data in '*.spec.tsx' test file and passing it to the tested component in ```mount```, initial render works.
However, upon updating the ```ref``` value with a new value, the component doesn't automatically rerender.

This is, for example, working in Cypress component testing (mentioning it since I just migrated to Playwright).

I understand one can use Playwright's ```await component.rerender({ props: { *somepropname*: *someref.value* }})```

But that has a few significant drawbacks:
- It is not type safe, so upon refactoring component prop names, changing types etc., I will realize at best only after running the test. With reactivity, everything would be beautifully type safe and automatic.
- It doesn't really allow you to test deep reactivity (list item property changes etc.). Only shallowRef at best. You can force the rerender, yes, but that destroys the purpose of deep reactivity testing.

I admit that reactivity testing is not always useful, but I have some use for it in my more complex components.
More importantly, the type safety is just really really nice to have.
It is also part of a reason why I use tsx tests instead of ts tests; because only ```mount(<ItemList items={...} />))``` is type safe while also forcing you to provide values for all required props and warning you when providing values for non-existent props.

**Full repro project. Steps to reproduce:**
- Download [repro-project.zip](https://github.com/microsoft/playwright/files/9878311/repro-project.zip)
- Install *Volar* extension (official Vue 3 support extension)
- Enable Take Over mode for workspace (guide is in *README.md* if needed)
- ```npm install```
- ```npm run dev``` to see a glorious ItemList.vue component drawing ten items
- Check the 'ItemList.vue' component code
- Check the 'ItemList.spec.tsx' test code
- Try running the test from previous bullet. The test fails. There are a few exaplanatory comments in the test file with additional info.

**Minor additional issue also demonstrated in the repro project**:
```interface MountResult``` from ```@playwright/experimental-ct-vue``` is not exported.
This is a little annoying since I can only pass the component reference as ```Locator``` from ```beforeEach``` where I'm mounting the actual component.
Then, in the aforementioned test file I have to do this ```await (component as any).rerender(...)```.

**Comment by sand4rt** [comment]
> It is not type safe, so upon refactoring component prop names, changing types etc., I will realize at best only after running the test. With reactivity, everything would be beautifully type safe and automatic.

The renderer function has had some improvements in the meantime and It will be typed in a future version. It's still in alpha stage, but if you want to check that out already, you can use: `"@playwright/experimental-ct-vue": "^1.28.0-alpha-oct-27-2022",`. Note that at a later stage the `rerender` will probably be renamed to `update`: https://github.com/microsoft/playwright/blob/904a3b46e3d717b88b3fbbc787d36b0f6bb2ff70/tests/components/ct-react-vite/src/tests.spec.tsx#L22-L31

> It doesn't really allow you to test deep reactivity (list item property changes etc.). Only shallowRef at best. You can force the rerender, yes, but that destroys the purpose of deep reactivity testing.

`rerender`/`update` doesn't destroy and recreate the Vue instance so it should suffice to test the reactivity.

> The type safety is just really really nice to have. It is also part of a reason why I use tsx tests instead of ts tests

Yeah definitely! The ts/object API may also be typed in the future: https://github.com/microsoft/playwright/issues/17206

**Comment by StepanMynarik** [comment]
> The renderer function has had some improvements in the meantime and It will be typed in a future version. It's still in alpha stage, but if you want to check that out already, you can use: "@playwright/experimental-ct-vue": "^1.28.0-alpha-oct-27-2022",. Note that at a later stage the rerender will probably be renamed to update:

Cool! Went straight ahead to try it.
Seems to be working, but this notation ```await components.rerender(<BMultiSelect label="" items={mockedData.items.value} modelValue={mockedData.selectedItems.value} />);``` always requires me to respecify all required parameters, even though I want ot update just one.

This is why refs in tests are so cool. It just behaves the same as in the actual app. In combination with TSX mounting it's just very useful and easy to use.

> rerender/update doesn't destroy and recreate the Vue instance so it should suffice to test the reactivity. I can't think of a good use case why you would want to create refs in the tests, then pass it to your component as props to update the refs.

On second though you are probably correct that deep reactivity testing is BS. I was using it mainly because of one very weird behaving 3rd party component. But I can live without it.
Still, refs for stringly typed updating are everything one might ever need.
I was using it with great success in Cypress and would miss it (I will not miss Cypress, it was driving me nuts, but I will definitely miss refs in tests).

> Yeah definitely! The ts/object API may also be typed in the future: https://github.com/microsoft/playwright/issues/17206

Well this would solve everything. If one could use it for type, required and non-existent prop validation on mount and type and non-existent prop validation on rerender.
That would be perfect!

**Comment by sand4rt** [comment]
> Cool! Went straight ahead to try it.
Seems to be working, but this notation `await components.rerender(<BMultiSelect label="" items={mockedData.items.value} modelValue={mockedData.selectedItems.value} />);` always requires me to respecify all required parameters, even though I want ot update just one.

Great! I agree that it is a minor inconvenience that it is necessary to specify all required props. In contrast, the need to create refs and pass them to the component and modify those refs is even more code to write and harder to read. In addition, with `rerender`/`update`, the API of the other frameworks can remain equal.

> Also please don't forget to export MountResult and MountResultJsx interfaces <3
IMO exposing `MountResult` leads to overly complicated tests. I would instead restructure the test a different way:

1. No need to generate a list of items, just a few should be enough
2. Use `rerender`/`update` instead of updating the `ref`
3. It is not necessary to always put the test in a `describe` block

```tsx
// ItemList.spec.tsx
import { expect, test } from '@playwright/experimental-ct-vue';
import ItemList from './ItemList.vue';
import Item from './item';

test('draw items', async ({ mount }) => {
  const component = await mount(<ItemList items={[
    new Item(1, 'Item 1'),
    new Item(2, 'Item 2'),
    new Item(3, 'Item 3'),
  ]} />);
  await expect(component.locator('.item')).toHaveCount(3);
});

test("update based on Vue reactivity", async ({ mount }) => {
  const component = await mount(<ItemList />);
  await component.rerender({
    props: {
      items: [new Item(1, 'Item 1'), new Item(2, 'Item 2')]
    }
  });
  await expect(component.locator('.item')).toHaveCount(2);
});
```

```tsx
// ItemList.vue
<script setup lang="ts">
import type Item from './item';

const { items = [] } = defineProps<{ items?: Item[] }>();
</script>

<template>
  <div class="vertical-list">
    <div class="item" v-for="item in items" :key="item.key">
      {{ item.label }}
    </div>
  </div>
</template>
```

**Comment by StepanMynarik** [comment]
> MO exposing MountResult leads to overly complicated tests.

This was but a subset of my tests. The component in my actual project is UI virtualized and for various tests I would keep recreating the same large set of items when I can do so in one place only.

> IMO exposing MountResult leads to overly complicated tests. I would instead reassemble the test a different way

You keep making suggestions on how I should implement my components, structure my code and tests.
This is fine, I like every opportunity to learn.
But testing framework should be flexible I believe.
It should enable me to do TDD, but also add it into a mature project to make sure perhaps just one overengineered piece of code doesn't misbehave without having to refactor my codebase based on someone elses opinions.

It should be possible to use with classes, types and interfaces "no matter where I put them".

These are my opinions, yes, but the workflow of schooling people on how they should write their tests and components is going to be exhausting eventually. Maybe I managed to annoy you already, and that is just me, one guy.

Just let me mount in separate method like foreach and then use rerender and perhaps other methods like unmount in other tests and callbacks.

Maybe I'm dumb, but I just like flexibility.

Minor detail: the second mount from your example makes VSCode scream at me for not specifying the required items prop.

I digress, sorry for that.
Rerender is fine. Specially if the 'rerender(ItemList, { props: { items: someList } })' will be typesafe. I could then update just the one prop.
This is something Cypress doesn't have and what forced me to use refs initially.
I agree with you about tests without refs being cleaner.
I would jus really like to have the option of type safely updating one or two props for example.

**Comment by sand4rt** [comment]
> This was but a subset of my tests. The component in my actual project is UI virtualized and for various tests I would keep recreating the same large set of items when I can do so in one place only.

> These are my opinions, yes, but the workflow of schooling people on how they should write their tests and components is going to be exhausting eventually. Maybe I managed to annoy you already, and that is just me, one guy

Let's wait for one of the core members _(note that i'm not the decision maker)_, but in my opinion that problem is separate from exposing `MountResult`. You could still create a function that generates mock data without having to call `mount` in a `beforeEach`. 

In my opinion, exposing the `MountResult` will lead to overly complicated tests. There are several other ways to reduce duplicate setup code. If you are interested see:
1. https://github.com/kentcdodds/kentcdodds.com/blob/main/content/blog/test-isolation-with-react.mdx
2. https://github.com/kentcdodds/kentcdodds.com/blob/main/content/blog/write-fewer-longer-tests/index.mdx

I think documentation on how to properly solve this problem, which we can also refer people to and better error messages, would be better than sending people down the wrong path.

> It should enable me to do TDD, but also add it into a mature project to make sure perhaps just one overengineered piece of code doesn't misbehave without having to refactor my codebase based on someone elses opinions. It should be possible to use with classes, types and interfaces "no matter where I put them".

Yeah for sure. You are able to use classes, types and interfaces. The limitation as of now is that you can't import it from the component file. I've just suggested some ideas to get around that for now.

> Minor detail: the second mount from your example makes VSCode scream at me for not specifying the required items prop.

See line: `const { items = [] } = defineProps<{ items?: Item[] }>();` to fix this

> I digress, sorry for that.
Rerender is fine. Specially if the 'rerender(ItemList, { props: { items: someList } })' will be typesafe. I could then update just the one prop.
This is something Cypress doesn't have and what forced me to use refs initially.
I agree with you about tests without refs being cleaner.
I would jus really like to have the option of type safely updating one or two props for example.

It's all right! Glad to see someone sharing their thoughts and having an open discussion to see where things _(on all sides)_ could improve! You certainly made some valid arguments.

**Comment by StepanMynarik** [comment]
> In my opinion, exposing the MountResult will lead to overly complicated tests. There are several other ways to reduce duplicate setup code. If you are interested see:
Looks cool, fair point. I will definitely give this a try.

> Yeah for sure. You are able to use classes, types and interfaces. The limitation as of now is that you can't import it from the component file. I've just suggested some ideas to get around that for now.
If it's considered a temporary limitation that should be alleviated in the future, then I'm absolutely happy :)
I was just reluctant to accept it as an intended permanent state of things.

> See line: const { items = [] } = defineProps<{ items?: Item[] }>(); to fix this
This is what I was talking about - I don't want to make workarounds in my "business logic" to make way for testing.
This parameter should stay required.
And generally I just use required parameters a lot. It is a great way to let users of your components know which parameters they must provide values for at minimum to render properly.

> That is alright! Glad to see someone sharing their thoughts and having an open discussion to see where things (on all sides) could improve! You certainly made some valid arguments.
It is a pleasure communicating with you :)

**Comment by sand4rt** [comment]
> This is what I was talking about - I don't want to make workarounds in my "business logic" to make way for testing.
This parameter should stay required.
And generally I just use required parameters a lot. It is a great way to let users of your components know which parameters they must provide values for at minimum to render properly.

Totally agree with that! Besides for the [import from a component file limitation](https://playwright.dev/docs/next/test-components#q-i-cant-import-anything-other-than-the-components-from-tsxjsxcomponent-files), you don't have to write the production code differently. I assumed you wanted it to be optional, because `items` has a default value if nothing is passed.

So instead of:
`const { items = [] } = defineProps<{ items?: Item[] }>();` 
it could be: 
`const { items } = defineProps<{ items: Item[] }>();`.
and the test could look something like this: 
```tsx
test("update based on Vue reactivity", async ({ mount }) => {
  const component = await mount(<ItemList items={[]} />);
  await component.rerender({
    props: {
      items: [new Item(1, 'Item 1'), new Item(2, 'Item 2')]
    }
  });
  await expect(component.locator('.item')).toHaveCount(2);
});
```

**Comment by pavelfeldman** [comment]
> If it's considered a temporary limitation that should be alleviated in the future, then I'm absolutely happy :) I was just reluctant to accept it as an intended permanent state of things.

I'll merge this issue into #18346 since I'm not sure there are other actionable items remaining in here.

**Comment by StepanMynarik** [comment]
I guess so, yes.
Just waiting for ```component.rerender``` to become type safe :)
